presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java

```diff
@@ -62,15 +64,17 @@ public SerializedPage serialize(Page page)
         SliceOutput serializationBuffer = new DynamicSliceOutput(toIntExact(page.getSizeInBytes() + Integer.BYTES)); // block length is an int
         writeRawPage(page, serializationBuffer, blockEncodingSerde);
         Slice slice = serializationBuffer.slice();
+
         int uncompressedSize = serializationBuffer.size();
         byte markers = PageCodecMarker.none();
 
         if (compressor.isPresent()) {
-            ByteBuffer compressionBuffer = ByteBuffer.allocate(compressor.get().maxCompressedLength(uncompressedSize));
-            compressor.get().compress(slice.toByteBuffer(), compressionBuffer);
-            compressionBuffer.flip();
-            if ((((double) compressionBuffer.remaining()) / uncompressedSize) <= MINIMUM_COMPRESSION_RATIO) {
-                slice = Slices.wrappedBuffer(compressionBuffer);
+            int maxCompressedSize = compressor.get().maxCompressedLength(uncompressedSize);
+            ensureCompressionBufferCapacity(maxCompressedSize);
+            int compressedSize = compressor.get().compress((byte[]) slice.getBase(), 0, uncompressedSize, compressionBuffer, 0, maxCompressedSize);
```

Comment:
While using `0` as the offset may be safe in this particular context, you might consider changing it to [this code](https://github.com/prestosql/presto/blob/ef67673a94a57212a540f922ab3d45116d599ed4/presto-main/src/main/java/io/prestosql/execution/buffer/PagesSerde.java#L72) so it's more resilient to future changes or someone copy/pasting this code without understanding it.